diff --git a/src/hotspot/share/classfile/javaClasses.cpp b/src/hotspot/share/classfile/javaClasses.cpp
index 3400b1d52..283ec307b 100644
--- a/src/hotspot/share/classfile/javaClasses.cpp
+++ b/src/hotspot/share/classfile/javaClasses.cpp
@@ -172,7 +172,7 @@ int java_lang_String::value_offset  = 0;
 int java_lang_String::hash_offset   = 0;
 int java_lang_String::coder_offset  = 0;
 
-bool java_lang_String::initialized  = false;
+int java_lang_String::initialized  = false;
 
 bool java_lang_String::is_instance(oop obj) {
   return is_instance_inlined(obj);
@@ -1521,7 +1521,7 @@ oop java_lang_Class::primitive_mirror(BasicType t) {
   return mirror;
 }
 
-bool java_lang_Class::offsets_computed = false;
+int  java_lang_Class::offsets_computed = false;
 int  java_lang_Class::classRedefinedCount_offset = -1;
 
 #define CLASS_FIELDS_DO(macro) \
@@ -4028,7 +4028,7 @@ oop java_security_AccessControlContext::create(objArrayHandle context, bool isPr
 
 // Support for java_lang_ClassLoader
 
-bool java_lang_ClassLoader::offsets_computed = false;
+int  java_lang_ClassLoader::offsets_computed = false;
 int  java_lang_ClassLoader::_loader_data_offset = -1;
 int  java_lang_ClassLoader::parallelCapable_offset = -1;
 int  java_lang_ClassLoader::name_offset = -1;
diff --git a/src/hotspot/share/classfile/javaClasses.hpp b/src/hotspot/share/classfile/javaClasses.hpp
index a982cecd5..ef898aa98 100644
--- a/src/hotspot/share/classfile/javaClasses.hpp
+++ b/src/hotspot/share/classfile/javaClasses.hpp
@@ -56,7 +56,7 @@ class java_lang_String : AllStatic {
   static int hash_offset;
   static int coder_offset;
 
-  static bool initialized;
+  static int initialized;
 
   static Handle basic_create(int length, bool byte_arr, TRAPS);
 
@@ -204,7 +204,7 @@ class java_lang_Class : AllStatic {
   static int _name_offset;
   static int _source_file_offset;
 
-  static bool offsets_computed;
+  static int offsets_computed;
   static int classRedefinedCount_offset;
 
   static GrowableArray<Klass*>* _fixup_mirror_list;
@@ -1276,7 +1276,7 @@ class java_security_AccessControlContext: AllStatic {
 class java_lang_ClassLoader : AllStatic {
  private:
   static int _loader_data_offset;
-  static bool offsets_computed;
+  static int offsets_computed;
   static int parent_offset;
   static int parallelCapable_offset;
   static int name_offset;
