--- jdk/src/hotspot/share/memory/metaspaceShared.cpp	2019-06-06 14:23:36.000000000 +0200
+++ new/src/hotspot/share/memory/metaspaceShared.cpp	2019-06-12 15:28:08.423443309 +0200
@@ -143,7 +143,7 @@
 public:
   DumpRegion(const char* name) : _name(name), _base(NULL), _top(NULL), _end(NULL), _is_packed(false) {}
 
-  char* allocate(size_t num_bytes, size_t alignment=BytesPerWord) {
+  char* allocate(size_t num_bytes, size_t alignment=BytesPerLong) {
     char* p = (char*)align_up(_top, alignment);
     char* newtop = p + align_up(num_bytes, alignment);
     expand_top_to(newtop);
@@ -1139,7 +1139,7 @@
     address obj = ref->obj();
     int bytes = ref->size() * BytesPerWord;
     char* p;
-    size_t alignment = BytesPerWord;
+    size_t alignment = BytesPerLong;
     char* oldtop;
     char* newtop;
 
@@ -1157,7 +1157,7 @@
         Klass* klass = (Klass*)obj;
         if (klass->is_instance_klass()) {
           SystemDictionaryShared::validate_before_archiving(InstanceKlass::cast(klass));
-          _rw_region.allocate(sizeof(address), BytesPerWord);
+          _rw_region.allocate(sizeof(u8), alignment);
         }
       }
       p = _rw_region.allocate(bytes, alignment);
