--- jdk/make/RunTests.gmk	2019-04-01 19:26:25.000000000 +0200
+++ new/make/RunTests.gmk	2019-06-12 10:54:26.919968395 +0200
@@ -844,6 +844,7 @@
   endif
 
   $1_JTREG_BASIC_OPTIONS += -e:TEST_IMAGE_GRAAL_DIR=${TEST_IMAGE_DIR}/hotspot/jtreg/graal
+  $1_JTREG_BASIC_OPTIONS += -e:ASAN_OPTIONS="handle_segv=0 detect_leaks=0"
 
   ifneq ($$(JTREG_FAILURE_HANDLER_OPTIONS), )
     $1_JTREG_LAUNCHER_OPTIONS += -Djava.library.path="$(JTREG_FAILURE_HANDLER_DIR)"
