diff --git a/src/hotspot/share/memory/metaspaceShared.cpp b/src/hotspot/share/memory/metaspaceShared.cpp
index 8ccc83a1c..3b6b73a9a 100644
--- a/src/hotspot/share/memory/metaspaceShared.cpp
+++ b/src/hotspot/share/memory/metaspaceShared.cpp
@@ -1205,7 +1205,7 @@ public:
     address obj = ref->obj();
     int bytes = ref->size() * BytesPerWord;
     char* p;
-    size_t alignment = BytesPerWord;
+    size_t alignment = BytesPerLong;
     char* oldtop;
     char* newtop;
 
@@ -1223,7 +1223,7 @@ public:
         Klass* klass = (Klass*)obj;
         if (klass->is_instance_klass()) {
           SystemDictionaryShared::validate_before_archiving(InstanceKlass::cast(klass));
-          _rw_region.allocate(sizeof(address), BytesPerWord);
+          _rw_region.allocate(sizeof(u8), alignment);
         }
       }
       p = _rw_region.allocate(bytes, alignment);
diff --git a/src/hotspot/share/memory/metaspaceShared.hpp b/src/hotspot/share/memory/metaspaceShared.hpp
index e293f60eb..687376403 100644
--- a/src/hotspot/share/memory/metaspaceShared.hpp
+++ b/src/hotspot/share/memory/metaspaceShared.hpp
@@ -67,7 +67,7 @@ public:
   DumpRegion(const char* name) : _name(name), _base(NULL), _top(NULL), _end(NULL), _is_packed(false) {}
 
   char* expand_top_to(char* newtop);
-  char* allocate(size_t num_bytes, size_t alignment=BytesPerWord);
+  char* allocate(size_t num_bytes, size_t alignment=BytesPerLong);
 
   void append_intptr_t(intptr_t n, bool need_to_mark = false);
 
