diff --git a/common/autoconf/flags.m4 b/common/autoconf/flags.m4
index a96b74be7..aaa3de469 100644
--- a/common/autoconf/flags.m4
+++ b/common/autoconf/flags.m4
@@ -30,7 +30,7 @@
 AC_DEFUN([FLAGS_SETUP_ABI_PROFILE],
 [
   AC_ARG_WITH(abi-profile, [AS_HELP_STRING([--with-abi-profile],
-      [specify ABI profile for ARM builds (arm-vfp-sflt,arm-vfp-hflt,arm-sflt, armv5-vfp-sflt,armv6-vfp-hflt,arm64,aarch64) @<:@toolchain dependent@:>@ ])])
+      [specify ABI profile for ARM builds (arm-vfp-sflt,arm-vfp-hflt,arm-sflt, arm926ejs,armv5-vfp-sflt,armv6-vfp-hflt,arm64,aarch64) @<:@toolchain dependent@:>@ ])])
 
   if test "x$with_abi_profile" != x; then
     if test "x$OPENJDK_TARGET_CPU" != xarm && \
@@ -51,6 +51,18 @@ AC_DEFUN([FLAGS_SETUP_ABI_PROFILE],
     elif test "x$OPENJDK_TARGET_ABI_PROFILE" = xarm-sflt; then
       ARM_FLOAT_TYPE=sflt
       ARM_ARCH_TYPE_FLAGS='-march=armv5t -marm'
+    elif test "x$OPENJDK_TARGET_ABI_PROFILE" = xarm926ejs; then
+      ARM_FLOAT_TYPE=sflt
+      ARM_ARCH_TYPE_FLAGS='-mcpu=arm926ej-s -marm'
     elif test "x$OPENJDK_TARGET_ABI_PROFILE" = xarmv5-vfp-sflt; then
       ARM_FLOAT_TYPE=vfp-sflt
       ARM_ARCH_TYPE_FLAGS='-march=armv5t -marm'
@@ -1305,6 +1317,14 @@ $2LDFLAGS_JDKLIB="${$2LDFLAGS_JDKLIB} ${$2JAVA_BASE_LDFLAGS}"
   if test "x$OPENJDK_$1_OS" = xlinux; then
     if test "x$OPENJDK_$1_CPU" = xx86; then
       $2JVM_ASFLAGS="[$]$2JVM_ASFLAGS -march=i586"
+    elif test "x$OPENJDK_TARGET_ABI_PROFILE" = xarm926ejs; then
+      $2JVM_ASFLAGS="-mcpu=arm926ej-s"
     fi
   elif test "x$OPENJDK_$1_OS" = xmacosx; then
     $2JVM_ASFLAGS="[$]$2JVM_ASFLAGS -x assembler-with-cpp -mno-omit-leaf-frame-pointer -mstack-alignment=16"
diff --git a/common/autoconf/lib-softfloat.m4 b/common/autoconf/lib-softfloat.m4
new file mode 100644
index 000000000..7fa1ed1ac
--- /dev/null
+++ b/common/autoconf/lib-softfloat.m4
@@ -0,0 +1,57 @@
+#
+# Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+#
+# This code is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 2 only, as
+# published by the Free Software Foundation.  Oracle designates this
+# particular file as subject to the "Classpath" exception as provided
+# by Oracle in the LICENSE file that accompanied this code.
+#
+# This code is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# version 2 for more details (a copy is included in the LICENSE file that
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+# or visit www.oracle.com if you need additional information or have any
+# questions.
+#
+
+################################################################################
+# Setup softfloat helper
+################################################################################
+AC_DEFUN_ONCE([LIB_SETUP_SOFTFLOAT],
+[
+  AC_ARG_WITH(softfloat, [AS_HELP_STRING([--with-softfloat],
+      [specify path to patched softfloat.a])])
+
+  if test "x$NEEDS_LIB_SOFTFLOAT" = xfalse; then
+    if (test "x${with_softfloat}" != x && test "x${with_softfloat}" != xno); then
+      AC_MSG_WARN([[softfloat not used, so --with-softfloat is ignored]])
+    fi
+    SOFTFLOAT_LIBS=
+  else
+    SOFTFLOAT_FOUND=no
+
+    if test "x${with_softfloat}" = xno; then
+      AC_MSG_ERROR([It is not possible to disable the use of softfloat. Remove the --without-softfloat option.])
+    fi
+
+    if test "x${with_softfloat}" != x; then
+      SOFTFLOAT_LIBS="${with_softfloat}"
+      SOFTFLOAT_FOUND=yes
+    fi
+    if test "x$SOFTFLOAT_FOUND" = xno; then
+      HELP_MSG_MISSING_DEPENDENCY([softfloat])
+      AC_MSG_ERROR([Could not find softfloat! $HELP_MSG])
+    fi
+  fi
+
+  AC_SUBST(SOFTFLOAT_LIBS)
+])
diff --git a/common/autoconf/libraries.m4 b/common/autoconf/libraries.m4
index bd5a40a7e..59157bb05 100644
--- a/common/autoconf/libraries.m4
+++ b/common/autoconf/libraries.m4
@@ -32,6 +32,7 @@ m4_include([lib-freetype.m4])
 m4_include([lib-std.m4])
 m4_include([lib-x11.m4])
 m4_include([lib-elf.m4])
+m4_include([lib-softfloat.m4])
 
 ################################################################################
 # Determine which libraries are needed for this configuration
@@ -68,6 +69,13 @@ AC_DEFUN_ONCE([LIB_DETERMINE_DEPENDENCIES],
     NEEDS_LIB_ALSA=false
   fi
 
+  if (test "x$OPENJDK_TARGET_CPU" == xarm &&
+      test "x$ARM_FLOAT_TYPE" = xsflt); then
+    NEEDS_LIB_SOFTFLOAT=true
+  else
+    NEEDS_LIB_SOFTFLOAT=false
+  fi
+
   # Check if ffi is needed
   if HOTSPOT_CHECK_JVM_VARIANT(zero) || HOTSPOT_CHECK_JVM_VARIANT(zeroshark); then
     NEEDS_LIB_FFI=true
@@ -86,6 +94,7 @@ AC_DEFUN_ONCE([LIB_SETUP_LIBRARIES],
   LIB_SETUP_CUPS
   LIB_SETUP_FREETYPE
   LIB_SETUP_ALSA
+  LIB_SETUP_SOFTFLOAT
   LIB_SETUP_LIBFFI
   LIB_SETUP_LLVM
   LIB_SETUP_BUNDLED_LIBS
diff --git a/common/autoconf/spec.gmk.in b/common/autoconf/spec.gmk.in
index aa86532e2..3911a8a6d 100644
--- a/common/autoconf/spec.gmk.in
+++ b/common/autoconf/spec.gmk.in
@@ -316,6 +316,7 @@ FREETYPE_LICENSE=@FREETYPE_LICENSE@
 CUPS_CFLAGS:=@CUPS_CFLAGS@
 ALSA_LIBS:=@ALSA_LIBS@
 ALSA_CFLAGS:=@ALSA_CFLAGS@
+SOFTFLOAT_LIBS:=@SOFTFLOAT_LIBS@
 LIBFFI_LIBS:=@LIBFFI_LIBS@
 LIBFFI_CFLAGS:=@LIBFFI_CFLAGS@
 ENABLE_LIBFFI_BUNDLING:=@ENABLE_LIBFFI_BUNDLING@
diff --git a/hotspot/make/lib/CompileJvm.gmk b/hotspot/make/lib/CompileJvm.gmk
index 4eb2cbc10..27640cdf9 100644
--- a/hotspot/make/lib/CompileJvm.gmk
+++ b/hotspot/make/lib/CompileJvm.gmk
@@ -107,6 +107,7 @@ JVM_LDFLAGS += \
 
 JVM_LIBS += \
     $(JVM_LIBS_FEATURES) \
+    $(SOFTFLOAT_LIBS) \
     #
 
 # These files and directories are always excluded
diff --git a/hotspot/src/cpu/arm/vm/assembler_arm_32.hpp b/hotspot/src/cpu/arm/vm/assembler_arm_32.hpp
index e32f6a98f..12c7d25e4 100644
--- a/hotspot/src/cpu/arm/vm/assembler_arm_32.hpp
+++ b/hotspot/src/cpu/arm/vm/assembler_arm_32.hpp
@@ -498,7 +498,7 @@ class Assembler : public AbstractAssembler  {
   void dmb(DMB_Opt opt, Register reg) {
     if (VM_Version::arm_arch() >= 7) {
       emit_int32(0xF57FF050 | opt);
-    } else {
+    } else if (VM_Version::arm_arch() == 6) {
       bool preserve_tmp = (reg == noreg);
       if(preserve_tmp) {
         reg = Rtemp;
@@ -1234,10 +1234,11 @@ extern int  __aeabi_dcmpgt(double, double);
 
 // Imported code from glibc soft-fp bundle for
 // calculation accuracy improvement. See CR 6757269.
-extern double __aeabi_fadd_glibc(float, float);
-extern double __aeabi_fsub_glibc(float, float);
+extern float  __aeabi_fadd_glibc(float, float);
+extern float  __aeabi_fsub_glibc(float, float);
 extern double __aeabi_dadd_glibc(double, double);
 extern double __aeabi_dsub_glibc(double, double);
+
 };
 #endif // __SOFTFP__
 
diff --git a/hotspot/src/cpu/arm/vm/vm_version_arm_32.cpp b/hotspot/src/cpu/arm/vm/vm_version_arm_32.cpp
index 10a240877..53223087e 100644
--- a/hotspot/src/cpu/arm/vm/vm_version_arm_32.cpp
+++ b/hotspot/src/cpu/arm/vm/vm_version_arm_32.cpp
@@ -295,6 +295,15 @@ void VM_Version::initialize() {
     FLAG_SET_DEFAULT(UseUnalignedAccesses, false);
   }
 
+  // ARMv5 MP workaround
+  if (arm_arch() == 5) {
+    if (FLAG_IS_DEFAULT(AssumeMP)) {
+      FLAG_SET_DEFAULT(AssumeMP, false);
+    } else if (AssumeMP) {
+      warning("AssumeMP can not be true for ARMv5 as there is only uniprocessor support");
+      FLAG_SET_DEFAULT(AssumeMP, false);
+    }
+  }
   _is_initialized = true;
 }
 
diff --git a/hotspot/src/os_cpu/linux_arm/vm/orderAccess_linux_arm.inline.hpp b/hotspot/src/os_cpu/linux_arm/vm/orderAccess_linux_arm.inline.hpp
index 67e32a834..3627c8ea6 100644
--- a/hotspot/src/os_cpu/linux_arm/vm/orderAccess_linux_arm.inline.hpp
+++ b/hotspot/src/os_cpu/linux_arm/vm/orderAccess_linux_arm.inline.hpp
@@ -72,7 +72,7 @@ inline static void dmb_sy() {
      __asm__ volatile (
      ".word 0xF57FF050 | 0xf" : : : "memory");
 #endif
-   } else {
+   } else if (VM_Version::arm_arch() == 6) {
      intptr_t zero = 0;
      __asm__ volatile (
        "mcr p15, 0, %0, c7, c10, 5"
@@ -96,7 +96,7 @@ inline static void dmb_st() {
      __asm__ volatile (
      ".word 0xF57FF050 | 0xe" : : : "memory");
 #endif
-   } else {
+   } else if (VM_Version::arm_arch() == 6) {
      intptr_t zero = 0;
      __asm__ volatile (
        "mcr p15, 0, %0, c7, c10, 5"
